src/int.rs

```diff
@@ -0,0 +1,470 @@
+//! Stack-allocated big signed integers.
+
+use core::fmt;
+
+use num_traits::ConstZero;
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq, ConstantTimeGreater, CtOption};
+
+use crate::{
+    Bounded, ConstChoice, ConstCtOption, ConstantTimeSelect, Encoding, Limb, NonZero, Odd, Uint,
+    Word,
+};
+
+#[macro_use]
+mod macros;
+
+mod add;
+mod cmp;
+mod div;
+mod encoding;
+mod expand;
+mod mul;
+mod sub;
+
+#[cfg(feature = "rand_core")]
+mod rand;
+
+/// Stack-allocated big _signed_ integer.
+/// See [`Uint`] for _unsigned_ integers.
+///
+/// Created as a [`Uint`] newtype.
+#[allow(clippy::derived_hash_with_manual_eq)]
+#[derive(Copy, Clone, Hash)]
+pub struct Int<const LIMBS: usize>(Uint<LIMBS>);
+
+impl<const LIMBS: usize> Int<LIMBS> {
+    /// The value `0`.
+    pub const ZERO: Self = Self(Uint::ZERO); // Bit sequence (be): 0000....0000
+
+    /// The value `1`.
+    pub const ONE: Self = Self(Uint::ONE); // Bit sequence (be): 0000....0001
+
+    /// The value `-1`
+    pub const MINUS_ONE: Self = Self::FULL_MASK; // Bit sequence (be): 1111....1111
+
+    /// Smallest value this [`Int`] can express.
+    pub const MIN: Self = Self(Uint::MAX.bitxor(&Uint::MAX.shr(1u32))); // Bit sequence (be): 1000....0000
+
+    /// Maximum value this [`Int`] can express.
+    pub const MAX: Self = Self(Uint::MAX.shr(1u32)); // Bit sequence (be): 0111....1111
+
+    /// Bit mask for the sign bit of this [`Int`].
+    pub const SIGN_BIT_MASK: Self = Self::MIN; // Bit sequence (be): 1000....0000
+
+    /// All-one bit mask.
+    pub const FULL_MASK: Self = Self(Uint::MAX); // Bit sequence (be): 1111...1111
+
+    /// Total size of the represented integer in bits.
+    pub const BITS: u32 = Uint::<LIMBS>::BITS;
+
+    /// Total size of the represented integer in bytes.
+    pub const BYTES: usize = Uint::<LIMBS>::BYTES;
+
+    /// The number of limbs used on this platform.
+    pub const LIMBS: usize = LIMBS;
+
+    /// Const-friendly [`Int`] constructor.
+    pub const fn new(limbs: [Limb; LIMBS]) -> Self {
+        Self(Uint::new(limbs))
+    }
+
+    /// Const-friendly [`Int`] constructor.
+    /// Note: interprets the `value` as an `Int`. For a proper conversion,
+    /// see [`Int::new_from_sign_and_magnitude`].
+    pub const fn new_from_uint(value: Uint<LIMBS>) -> Self {
```

Comment:
The name is a bit tricky... this is effectively a casting/transmutation operation that reinterprets the bits of a `Uint` as an `Int`.

Reply:
yeah, I was having similar problems with naming and couldn't come up with anything better. you have `as` operator in rust that does this for primitive types so maybe we can mimic that and implement as<Uint> as<Int> (or cast<...>?) to be consistent with at least that.

Reply:
The only thing that comes to mind for me is [`from_bits`](https://doc.rust-lang.org/std/primitive.f32.html#method.from_bits) which reinterprets an integer as a floating point

Reply:
Perhaps [`transmute`](https://doc.rust-lang.org/std/mem/fn.transmute.html) is the way to go?

> Reinterprets the bits of a value of one type as another type.
> Both types must have the same size. Compilation will fail if this is not guaranteed.

 I is mentioned in the `from_bits` doc that it is identical to `transmute`.

Reply:
@lleoha wait, one cannot implement `as`, right? Because, I agree, this definitely sounds like an `as` operation. I would steer clear from using `from` / `into`, as those names suggests a "proper" conversion, which is definitely not the case here.

Reply:
@erik-3milabs Yes, you cannot overload `as` operator but I am big fan of consistency so somewhat see no reason why you shouldn't be able to do something like:
`Int::<2>::ONE.as_uint()` or `Int::<2>::ONE.as::<Uint>()`
when you can do `1_i128 as u128`.
I am also completely fine with `from_bits` and/or `transmute`.

Reply:
Ahh. That is a curious, but elegant solution! Never thought about that :thinking:

Reply:
I renamed `new_from_uint` to `pub(crate) const fn from_bits` and introduced `Uint::as_int`; `Int::as_uint` already existed.